In [0]:
import numpy as np

# Exercise on finding number of points outside n-dimensional sphere

In [0]:
ndim = 2

In [0]:
npoints = 1000000

In [0]:
points = np.random.rand(npoints, ndim)

In [22]:
points[0:2, :]

array([[0.56553273, 0.99586325],
       [0.94473961, 0.70479456]])

In [0]:
dfo = np.zeros((npoints, 1))
outside_points = 0

In [24]:
%%time
for i in range(npoints):
    for j in range(ndim):
        dfo[i] += points[i, j] ** 2
    dfo[i] = np.sqrt(dfo[i])
    if dfo[i] > 1:
        outside_points += 1

CPU times: user 10.6 s, sys: 0 ns, total: 10.6 s
Wall time: 10.6 s


In [25]:
print('Fraction of points outside is ', outside_points/npoints)

Fraction of points outside is  0.214931


In [26]:
# 1 - (pi/4)
1 - 3.14/4

0.21499999999999997

In [27]:
%%time
sq_points = points * points
dfo = np.sqrt(np.sum(sq_points, axis = 1))
outside_points = np.sum(dfo > 1)

CPU times: user 27.6 ms, sys: 10 ms, total: 37.7 ms
Wall time: 38 ms


In [28]:
print('Fraction of points outside is ', outside_points/npoints)

Fraction of points outside is  0.214931


In [29]:
%%time
outside_points = np.sum(np.sqrt(np.sum(points * points, axis = 1)) > 1)

CPU times: user 23.2 ms, sys: 2.97 ms, total: 26.2 ms
Wall time: 27.2 ms


In [0]:
def area_outside_square(npoints, ndim):
    points = np.random.rand(npoints, ndim)
    return np.sum(np.sqrt(np.sum(points * points, axis = 1)) > 1)/npoints

In [31]:
area_outside_square(100000, 2)

0.21727

In [32]:
for i in range(2, 11):
    print(i, area_outside_square(100000, i))

2 0.21675
3 0.4798
4 0.69299
5 0.83668
6 0.91936
7 0.96404
8 0.98377
9 0.99328
10 0.99729


# Case study Cricket

In [33]:
!head cric_data.tsv

	Sachin Tendulkar	Rahul Dravid	India
0	100	78	342
1	11	62	191
2	8	85	252
3	71	24	307
4	104	17	229
5	18	104	246
6	8	76	226
7	86	74	288
8	12	60	216




1.   Find mean, median, IQR for Sachin, Rahul, and India
2.   Find the histogram of Sachin's scores with 10 bins
3.   Find mean of Sachin's scores grouped by 25 matches
4.   Find mean of Sachin's scores where he has scored a century
5.   Find mean of Sachin's scores when Rahul has scored less than 10
6.   Find mean for Sachin's scores based on which quartile India's score falls in
7.   For every match find out who has has scored more - Sachin or Rahul
8.   How many more runs does Sachin score on average after having scored x runs
9.   How many matches did Sachin take to score first 1000 runs, next 1000 runs, ...



### Problem 1

In [34]:
cric_data = np.loadtxt("cric_data.tsv")

ValueError: ignored

In [0]:
cric_data = np.loadtxt("cric_data.tsv", skiprows=1)

In [36]:
cric_data.shape

(225, 4)

In [0]:
cric_data = cric_data[:, [1, 2, 3]]

In [38]:
cric_data.shape

(225, 3)

In [0]:
sachin = cric_data[:, 0]

In [0]:
rahul = cric_data[:, 1]

In [0]:
india = cric_data[:, 2]

In [0]:
def stats(col):
    print('Mean', np.mean(col))
    print('Median', np.median(col))
    print('IQR', np.percentile(col, 75) 
    - np.percentile(col, 25))

In [43]:
stats(sachin)

Mean 39.87555555555556
Median 27.0
IQR 57.0


In [44]:
stats(rahul)

Mean 32.062222222222225
Median 22.0
IQR 46.0


In [45]:
stats(india)

Mean 220.79555555555555
Median 216.0
IQR 98.0


In [46]:
np.mean(cric_data, axis = 0)

array([ 39.87555556,  32.06222222, 220.79555556])

In [47]:
np.median(cric_data, axis = 0)

array([ 27.,  22., 216.])

In [48]:
np.percentile(cric_data, 75, axis = 0) - np.percentile(cric_data, 25, axis = 0)

array([57., 46., 98.])

##Problem 2

In [49]:
np.histogram(sachin)

(array([99, 36, 28, 16, 11, 17,  8,  8,  1,  1]),
 array([  0. ,  18.6,  37.2,  55.8,  74.4,  93. , 111.6, 130.2, 148.8,
        167.4, 186. ]))

## Problem 3

In [50]:
sachin.shape

(225,)

In [51]:
sachin.reshape(9, 25).shape

(9, 25)

In [0]:
sachin_by_25s = sachin.reshape(9, 25)

In [53]:
np.mean(sachin_by_25s, axis = 1)

array([33.96, 49.4 , 38.48, 40.16, 39.36, 38.2 , 44.6 , 39.52, 35.2 ])

Problem 4, 5

In [54]:
np.mean(sachin[sachin >= 100])

125.0

In [55]:
np.mean(sachin[rahul <= 10])

40.2112676056338

## Problem 6

In [0]:
qrs = np.percentile(india, [25, 50, 75, 100])

If India <= 175: Sachin's average is ....

If India <= 216: Sachin's average is ....

<= 273
<= 499



In [57]:
india < qrs

ValueError: ignored

In [58]:
india.shape

(225,)

In [59]:
qrs.shape

(4,)

In [0]:
qrs = qrs.reshape(4, 1)

In [0]:
indices = india < qrs

In [62]:
indices.shape

(4, 225)

In [63]:
sachin[indices[1, :]]

array([ 11.,  18.,  14.,   4.,   0.,  21.,  62.,   2.,  46.,  65.,   0.,
        39.,  48.,   3.,  11.,   5.,  65.,  27.,   7.,   2.,  28.,   6.,
         7.,   3.,   2.,   4., 100.,  15.,  32.,  40.,   0.,   5.,   8.,
         5.,   0.,  30.,  37.,  89.,   0.,  52.,  93.,  45.,   0.,   1.,
         0.,   3.,   0.,  81.,  78.,   6.,  28.,  13.,  93.,   2.,  36.,
         2.,  12.,  19.,  34.,   2.,  82.,   0.,   6.,   8.,  23.,  35.,
        74., 117.,  49.,  64.,   5.,  17.,   0.,  44.,  68.,   3.,  47.,
        17.,  35.,  88., 127.,   0.,  45.,  33., 110.,   7.,  25.,   9.,
         2.,  11.,  17.,   1.,  45.,   9.,  67.,  10.,  55.,   0.,   1.,
       101.,  23.,   1.,  41.,   1.,  28.,   2.,  22.,  25.,   0.,  27.,
        40.])

In [64]:
for i in range(4):
    print(i, np.mean(sachin[indices[i]]))

0 19.672727272727272
1 28.18018018018018
2 31.688622754491018
3 39.799107142857146


## Problem 7

In [0]:
snr = cric_data[:, 0:2]

In [66]:
np.argmax([10, 3, 2, 5, 1])

0

In [0]:
is_rahul_higher = np.argmax(snr, axis=1)

In [68]:
np.sum(is_rahul_higher) / 225

0.4711111111111111

In [69]:
np.where(is_rahul_higher == 0, 'Sachin', 'Rahul')

array(['Sachin', 'Rahul', 'Rahul', 'Sachin', 'Sachin', 'Rahul', 'Rahul',
       'Sachin', 'Rahul', 'Sachin', 'Rahul', 'Rahul', 'Rahul', 'Sachin',
       'Sachin', 'Rahul', 'Rahul', 'Sachin', 'Rahul', 'Rahul', 'Sachin',
       'Rahul', 'Sachin', 'Sachin', 'Rahul', 'Sachin', 'Sachin', 'Sachin',
       'Sachin', 'Sachin', 'Sachin', 'Sachin', 'Rahul', 'Rahul', 'Sachin',
       'Rahul', 'Rahul', 'Sachin', 'Rahul', 'Sachin', 'Sachin', 'Sachin',
       'Sachin', 'Rahul', 'Sachin', 'Rahul', 'Rahul', 'Sachin', 'Rahul',
       'Sachin', 'Rahul', 'Rahul', 'Rahul', 'Rahul', 'Sachin', 'Rahul',
       'Sachin', 'Sachin', 'Rahul', 'Rahul', 'Rahul', 'Sachin', 'Rahul',
       'Sachin', 'Sachin', 'Sachin', 'Sachin', 'Rahul', 'Sachin', 'Rahul',
       'Rahul', 'Sachin', 'Sachin', 'Rahul', 'Rahul', 'Sachin', 'Rahul',
       'Rahul', 'Sachin', 'Rahul', 'Rahul', 'Rahul', 'Rahul', 'Rahul',
       'Rahul', 'Sachin', 'Sachin', 'Sachin', 'Rahul', 'Sachin', 'Sachin',
       'Sachin', 'Sachin', 'Sachin', 'Sachin'

## Problem 8

In [0]:
x_arr = np.arange(0, 101, 5)

In [71]:
sachin >= x_arr

ValueError: ignored

In [0]:
x_arr = x_arr.reshape(x_arr.shape[0], 1)

In [0]:
indices = (sachin >= x_arr)

In [74]:
indices.shape

(21, 225)

In [75]:
sachin[indices[1, :]]

array([100.,  11.,   8.,  71., 104.,  18.,   8.,  86.,  12.,  85.,  18.,
         7.,  37.,  14.,  21.,  62., 138.,  38.,  46.,  65.,  39.,  48.,
       141.,  62.,  12.,  41.,  11., 186.,  11.,  27.,  27.,  51.,  18.,
        32., 146.,   5.,  45., 141.,  12.,  65.,  27.,   7.,  16.,  28.,
         6., 123., 120.,   7.,  81.,  54., 122.,  14., 100.,  15.,  57.,
        99.,  37.,  38.,  32.,  21.,  32.,  40.,   5.,   8.,   5.,  50.,
        30.,  37.,  89.,  98.,  83.,  93.,  52., 152.,   8.,  93.,  45.,
        26.,  16.,  47.,  89.,  53.,  16.,  81.,  14.,  78.,   6., 105.,
       122.,   9.,   8.,  28.,  35.,  69.,  13.,  97.,  93.,  36.,  39.,
        29.,  12.,  19.,  34., 100.,  44.,  82.,  79.,   6.,   9.,   8.,
        23.,  93.,  35.,  63.,  74.,   8., 117.,  39.,  49.,  64.,  43.,
        72.,   5.,  17.,  65.,  20., 141.,  28.,  44.,  27.,  60.,  68.,
       139.,  31.,  44.,  47.,   6.,  17.,  35.,  88., 114.,   7., 127.,
        45.,  33., 110., 146.,   7.,  25.,   9.,  1

In [76]:
for i in range(x_arr.shape[0]):
    print(x_arr[i, 0], 
          np.mean(sachin[indices[i, :]]) - x_arr[i, 0])

0 39.87555555555556
5 45.61363636363637
10 47.48026315789474
15 47.45255474452555
20 46.824
25 44.10084033613445
30 45.13461538461539
35 43.24742268041237
40 44.05882352941177
45 43.41558441558442
50 43.98529411764706
55 42.317460317460316
60 38.67213114754098
65 37.654545454545456
70 37.08163265306122
75 34.347826086956516
80 30.75
85 28.650000000000006
90 27.400000000000006
95 26.433333333333337
100 25.0


## Problem 9

In [77]:
sachin

array([100.,  11.,   8.,  71., 104.,  18.,   8.,  86.,  12.,  85.,  18.,
         4.,   7.,  37.,  14.,   0.,   4.,   0.,  21.,   1.,  62.,   0.,
       138.,  38.,   2.,  46.,  65.,   0.,  39.,  48., 141.,  62.,  12.,
         1.,  41.,  11.,   3., 186.,  11.,  27.,  27.,  51.,  18.,  32.,
       146.,   5.,  45., 141.,  12.,  65.,  27.,   7.,  16.,   2.,  28.,
         6., 123., 120.,   7.,   3.,   0.,  81.,   2.,  54., 122.,   4.,
        14.,   0., 100.,  15.,   0.,  57.,  99.,  37.,  38.,  32.,  21.,
        32.,  40.,   0.,   5.,   8.,   5.,   0.,  50.,  30.,  37.,  89.,
         4.,  98.,  83.,  93.,   0.,  52., 152.,   1.,   8.,  93.,  45.,
        26.,   0.,   1.,   0.,  16.,  47.,  89.,   3.,   1.,  53.,  16.,
         0.,  81.,  14.,  78.,   6., 105., 122.,   9.,   8.,  28.,  35.,
        69.,  13.,  97.,  93.,   2.,  36.,  39.,   2.,  29.,  12.,  19.,
        34.,   2., 100.,  44.,  82.,   0.,  79.,   6.,   9.,   8.,  23.,
        93.,  35.,  63.,  74.,   8., 117.,  39.,  4

In [0]:
sachin_cumscores = np.cumsum(sachin)

In [79]:
np.histogram(sachin_cumscores, bins=np.arange(0, 10000, 1000))

(array([29, 18, 26, 25, 26, 26, 23, 22, 30]),
 array([   0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000]))